conda/common/logic.py

```diff
@@ -29,26 +29,195 @@
 """
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+from array import array
 from itertools import chain, combinations
-from logging import getLogger
+from logging import DEBUG, getLogger
 import pycosat
 
 from .compat import iteritems
+from .io import time_recorder
 
 log = getLogger(__name__)
 
 
+class ClauseList(object):
+    def __init__(self):
+        self._clause_list = []
+        self.append = self._clause_list.append
+        self.extend = self._clause_list.extend
+        self.get_clause_count = self._clause_list.__len__
+
+    def save_state(self):
+        return len(self._clause_list)
+
+    def restore_state(self, saved_state):
+        len_clauses = saved_state
+        self._clause_list[len_clauses:] = []
+
+    def as_list(self):
+        return self._clause_list
+
+    def as_array(self):
+        clause_array = array('i')
+        for c in self._clause_list:
+            clause_array.extend(c)
+            clause_array.append(0)
+        return clause_array
+
+
+class ClauseArray(object):
+    def __init__(self):
+        self._clause_array = array('i')
+        self._array_append = self._clause_array.append
+        self._array_extend = self._clause_array.extend
+
+    def extend(self, clauses):
+        for clause in clauses:
+            self.append(clause)
+
+    def append(self, clause):
+        self._array_extend(clause)
+        self._array_append(0)
+
+    def get_clause_count(self):
+        return self._clause_array.count(0)
```

Comment:
I would make sense to make this an `O(1)` operation just like `ClausesList.get_clause_count`, but due to CPython interpreter overhead, simple counting in `ClauseArray` has a slight performance impact :unamused: